torch/nested/_internal/nested_tensor.py

```diff
@@ -3,23 +3,224 @@
+    # in eager to access the cpu and cuda variants from one another. Note that
+    # that since each equiv set has its own cache, FakeTensor must be in a
+    # different equiv set from real tensor.
+    #
+    def __init__(self):
+        self._equiv_set_counter = 0
+        self._equiv_sets = WeakTensorKeyDictionary()
+        self._version_counters = WeakTensorKeyDictionary()
+
+    def contains_vec(self, vec):
+        return vec in self._equiv_sets
 
-def get_tensor_symint(tensor, *, coeff=1):
-    global _tensor_id_counter
-    tensor_symint = _tensor_symint_registry.get(tensor)
-    if tensor_symint is None:
-        tensor_symint = torch._C._get_nested_int(_tensor_id_counter, coeff)
-        _tensor_id_counter += 1
-        _tensor_symint_registry[tensor] = tensor_symint
-    return tensor_symint
+    def assert_contains_vec(self, vec):
+        assert self.contains_vec(vec), (
+            "Expected vec to have been registered. "
+        )
 
+    def check_version_counter(self, vec):
+        # Check that vec has not been mutated
+        assert self._version_counters[vec] == vec._version, (
+            "Detected that vec has been mutated. This is not allowed. "
+        )
+
+    def maybe_create(self, vec, *, ctor_fn=None, equiv_set_from=None, has_coeff=False):
+        # Given vec, return an associated nested int.
+        #
+        # Parameters:
+        #     ctor_fn (Callable[[int, Tensor], SymInt]): If not None, use a custom
+        #        constructor to create the nested int. A equiv set is "custom"
+        #        if its canonical int was created using a custom ctor_fn (This
+        #        is useful during compile). The custom-ness of a equiv set is
+        #        immutable. This is because (1) unlike ordinary equiv set,
+        #        canonical nested int of a custom equiv set cannot be changed
+        #        i.e., if the nested int of a custom equiv set dies, don't
+        #        allow re-creation, and (2) during recreation of canonical nested
+        #        int, ctor_fn cannot be used, so a non-custom equiv set cannot
+        #        become custom once created.
+        #     equiv_set_from (Tensor): If not None, add vec to the equiv set of
+        #        the equiv set corresponding to the vec specified by this arg.
+        #        The user is responsible for ensuring that vec and equiv_set_from
+        #        are compatible, i.e., they have the same data.
+        #     has_coeff (bool): If True, a new nested int will be created even
+        #        if vec is already in an equiv set. Note that we don't actually
+        #        allow coeff to passed to this function. The ctor_fn is expected
+        #        to handle this. Also note that this is not how nested ints with
+        #        coeff are typically created. Ordinarily, nested ints with coeff
+        #        are created by operations on existing nested ints.
```

Comment:
nit: arg name and behavior don't match to me intuitively

Reply:
Yeah, the behavior is quite general, and the arg name hints at the specific and only use case it is used for. Let me try to see if I can make this make more sense.